File: cs/soa-benchmark/soa-benchmark/Aor.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;

namespace soa_benchmark
{
    public class Aor
    {
        public List<ListItem> Items;
        public int Size;

        public Aor(int size)
        {
            Size = size;
            Items = new List<ListItem>(size);
            
            Fill();
        }

        public void Fill()
        {
            var r = new Random();
            for (var i = 0; i < Size; i++)
            {
                Items.Add(new ListItem(r.Next(), r.Next() %10 > 5, $"Item {i}", $"http://item.dev/{i}"));
            }
        }

        public int[] RandomIds()
        {
            var r = new Random();
            var lids = new int[Size];
            var rids = new int[Size];
            for (var i = 0; i < Size; i++)
            {
                lids[i] = Items[i].Id;
            }
            for (var i = 0; i < Size; i++)
            {
                rids[i] = lids[r.Next(0, Size)];
            }

            return rids;
        }

        public bool Selected(int id)
        {
            foreach (var item in Items)
            {
                if (item.Id == id)
                {
                    return item.Selected;
                }
            }

            return false;
        }
    }
}

File: scala/src/main/scala/Aor.scala
package dev.csn

import scala.collection.mutable.ArrayBuffer

class Aor(size: Int) {
  val items = ArrayBuffer[ListItem]()

  fill()

  def fill() = {
    val r = scala.util.Random
    for (_ <- 0 to size) do {
      val id = r.nextInt()
      items += ListItem(id, r.nextBoolean, s"Item ${id}", s"http://item.dev/${id}")
    }
  }

  def randomIds(): Array[Int] = {
    val r = scala.util.Random
    val lids = items.map(_.id).toArray
    val indexes = for _ <- 0 to size yield r.nextInt(size)
    return indexes.map(lids(_)).toArray
  }

  def selected(id: Int): Boolean =
    items.find(_.id == id).get.selected

}
